src/doc/trpl/closures.md

```diff
@@ -1,214 +1,510 @@
+works well.
 
-There's one big difference between a closure and named functions, and it's in
-the name: a closure "closes over its environment." What does that mean? It means
-this:
+The second is that the syntax is similar, but a bit different. I've added spaces
+here to make them look a little closer:
 
-```{rust}
+```rust
+fn  plus_1 (  x: i32 ) -> i32 { x + 1 }
+let plus_2 = |x: i32 | -> i32 { x + 1 };
+let plus_3 = |x      |          x + 1  ;
+```
+
+Small differences, but they're similar in ways.
+
+# Closures and their environment
+
+As we said, closures are called such because they 'close over their environment.' It
+looks like this:
+
+```rust
+let num = 5;
+let plus_num = |x| x + num;
+
+assert_eq!(10, plus_num(5));
+```
+
+This closure, `plus_num`, refers to a `let` binding in its scope: `num`. More
+specifically, it borrows the binding. If we do something that would conflict
+with that binding, we get an error. Like this one:
+
+```rust,ignore
+let mut num = 5;
+let plus_num = |x| x + num;
+
+let y = &mut num;
+```
+
+Which errors with:
+
+```text
+error: cannot borrow `num` as mutable because it is also borrowed as immutable
+    let y = &mut num;
+                 ^~~
+note: previous borrow of `num` occurs here due to use in closure; the immutable
+  borrow prevents subsequent moves or mutable borrows of `num` until the borrow
+  ends
+    let plus_num = |x| x + num;
+                   ^~~~~~~~~~~
+note: previous borrow ends here
 fn main() {
-    let x: i32 = 5;
+    let mut num = 5;
+    let plus_num = |x| x + num;
+    
+    let y = &mut num;
+}
+^
+```
+
+A verbose yet helpful error message! As it says, we can't take a mutable borrow
+on `num` because the closure is already borrowing it. If we let the closure go
+out of scope, we can:
+
+```rust
+let mut num = 5;
+{
+    let plus_num = |x| x + num;
+
+} // plus_num goes out of scope, borrow of num ends
+
+let y = &mut num;
+```
+
+If your closure requires it, however, Rust will take ownership and move
+the environment instead:
 
-    let printer = || { println!("x is: {}", x); };
+```rust,ignore
+let nums = vec![1, 2, 3];
 
-    printer(); // prints "x is: 5"
+let takes_nums = || nums;
+
+println!("{:?}", nums);
+```
+
+This gives us:
+
+```text
+note: `nums` moved into closure environment here because it has type
+  `[closure(()) -> collections::vec::Vec<i32>]`, which is non-copyable
+let takes_nums = || nums;
+                    ^~~~~~~
+```
+
+`Vec<T>` has ownership over its contents, and therefore, when we refer to it
+in our closure, we have to take ownership of `nums`. It's the same as if we'd
+passed `nums` to a function that took ownership of it.
+
+## `move` closures
+
+We can force our closure to take ownership of its environment with the `move`
+keyword:
+
+```rust
+let num = 5;
+
+let owns_num = move |x| x + num;
+```
+
+Now, even though the keyword is `move`, the variables follow normal move semantics.
```

Comment:
Should it be like this? It's written like `move` is contrasted with something but it's not being contrasted.

> Now, ~~even though~~ the keyword `move` [implies] ~~, the variables follow~~ normal move semantics.